autogen/oai/gemini.py

```diff
@@ -342,23 +413,55 @@ def _oai_messages_to_gemini_messages(self, messages: list[Dict[str, Any]]) -> li
         rst = []
         curr_parts = []
         for i, message in enumerate(messages):
-            parts = self._oai_content_to_gemini_content(message["content"])
+
+            # Since the tool call message does not have the "name" field, we need to find the corresponding tool message.
+            if message["role"] == "tool":
+                message["name"] = [
+                    m for m in messages if "tool_calls" in m and m["tool_calls"][0]["id"] == message["tool_call_id"]
+                ][0]["tool_calls"][0]["function"]["name"]
+
+            parts = self._oai_content_to_gemini_content(message)
             role = "user" if message["role"] in ["user", "system"] else "model"
 
             if prev_role is None or role == prev_role:
-                curr_parts += parts
-            elif role != prev_role:
+                # If the message is a function call or a function response, we need to separate it from the previous message.
                 if self.use_vertexai:
-                    rst.append(VertexAIContent(parts=self._concat_parts(curr_parts), role=prev_role))
+                    if parts[0].function_call or parts[0].function_response:
+                        if len(curr_parts) > 1:
+                            rst.append(VertexAIContent(parts=self._concat_parts(curr_parts), role=prev_role))
+                        elif len(curr_parts) == 1:
+                            rst.append(VertexAIContent(parts=curr_parts, role=None if curr_parts[0].function_response else role))
+                        rst.append(VertexAIContent(parts=parts, role="user" if parts[0].function_response else role))
+                        rst.append(VertexAIContent(parts=self._oai_content_to_gemini_content("continue"), role="model"))
```

Comment:
@arjun-g this line needs to be removed, you don't want a message with role=model to be appended after a function response, otherwise you'll get this error from Google: `Please ensure that function call turn comes immediately after a user turn or after a function response turn.`

Reply:
Gemini requires a "function_response" message should be followed by "model" message. But this implementation was failing in some cases. Now changed the way this is appended.